### DISCUSSION TITLE
Example BAND cnbdr2.f

### DISCUSSION THREAD

**Discussion by wo80** [opencollab/arpack-ng#443] [Q&A] [answered]
This is not an issue, I'm just looking for confirmation.

The [BAND/cnbdr2.f](https://github.com/opencollab/arpack-ng/blob/master/EXAMPLES/BAND/cnbdr2.f) example exits with

    Error with _naupd info = -9 Check the documentation of _naupd

This means the initial residual was 0, which seems to make sense, since the example uses mode 3,
https://github.com/opencollab/arpack-ng/blob/569a3859c1bf79b303d830ada7b745de0f18512c/EXAMPLES/BAND/cnbdr2.f#L172
which accesses the M matrix
https://github.com/opencollab/arpack-ng/blob/569a3859c1bf79b303d830ada7b745de0f18512c/EXAMPLES/BAND/cnband.f#L492-L503
 but that is never set (just filled with zeros).

I'm running the examples as tests, so I just need someone to confirm that this is expected behavior. All the other examples run successfully (EDIT: well, znbdr2.f also fails).

**Comment by wo80** [comment, marked as answer]
So, since this example seems to solve a standard eigenvalue problem, setting M to the identity matrix should be enough (cnbdr2.f lines 199-209):
```fortran
c     %---------------%
c     | Main diagonal |
c     %---------------%
c
      h = one / cmplx(nxi+1)
      h2 = h*h
c
      idiag = kl+ku+1
      do 30 j = 1, n
         a(idiag,j) = (4.0E+0, 0.0E+0)  / h2
         m(idiag,j) = (1.0E+0, 0.0E+0)
  30  continue
```
Same for znbdr2.f. Error message gone...

**Reply by fghoussen** [reply]
On which OS do you see the `info = -9` error? Would make sense to PR this fix.

**Reply by wo80** [reply]
I checked on Debian 12 (with FlexiBLAS and OpenBLAS) and on Windows (msys2 with OpenBLAS).

**Comment by wo80** [comment]
I'd be glad to do a PR with the fix, but can someone else here confirm the issue? Would like to know if this is expected or if there's something wrong with my setup....

**Comment by wo80** [comment]
@fghoussen If the examples run correctly for you, could you please post their output, so I can compare with my results (with above fix applied).

cnbdr2
```
_NBDR2
======

 The size of the matrix is 100
 Number of eigenvalue requested is 4
 The number of Arnoldi vectors generated (NCV) is 10
 The number of converged Ritz values is 4
 What portion of the spectrum LM
 The number of Implicit Arnoldi update iterations taken is 9
 The number of OP*x is 53
 The convergence tolerance is 5.960464e-08


 Ritz values (Real,Imag) and relative residuals
 ----------------------------------------------
                Col    0        Col    1        Col    2
  Row    0:      2.51803e+02    1.52711e+02    6.55662e-07
  Row    1:      2.51803e+02   -1.52711e+02    1.07034e-06
  Row    2:      2.80417e+02    1.52711e+02    9.81432e-07
  Row    3:      2.80417e+02   -1.52711e+02    1.23237e-06
```

znbdr2
```
_NBDR2
======

 The size of the matrix is 100
 Number of eigenvalue requested is 4
 The number of Arnoldi vectors generated (NCV) is 10
 The number of converged Ritz values is 4
 What portion of the spectrum LM
 The number of Implicit Arnoldi update iterations taken is 19
 The number of OP*x is 105
 The convergence tolerance is 1.110223e-16


 Ritz values (Real,Imag) and relative residuals
 ----------------------------------------------
                Col    0        Col    1        Col    2
  Row    0:      2.51803e+02    1.52711e+02    2.35566e-15
  Row    1:      2.51803e+02   -1.52711e+02    2.80388e-15
  Row    2:      2.80417e+02    1.52711e+02    2.13067e-15
  Row    3:      2.80417e+02   -1.52711e+02    4.04126e-15
```

**Reply by fghoussen** [reply]
On debian x86_64 I get:

```
>> ./EXAMPLES/BAND/cnbdr2
  
  Error with _naupd info =           -9
  Check the documentation of _naupd 
  
  
  Error with _nband, info=           -9
  Check the documentation of _nband 

>> ./EXAMPLES/BAND/znbdr2
  
  Error with _naupd info =           -9
  Check the documentation of _naupd 
  
  
  Error with _nband, info=           -9
  Check the documentation of _nband 

```

The cmake non-reg test suite doesn't trap the error as it should...


```
>> make all test
>> sed -n -e 40,57p Testing/Temporary/LastTest.log 
2/77 Testing: cnbdr2_ex
2/77 Test: cnbdr2_ex
Command: "/tmp/arpack-ng/build/EXAMPLES/BAND/cnbdr2"
Directory: /tmp/arpack-ng/build
"cnbdr2_ex" start time: Nov 22 12:37 CET
Output:
----------------------------------------------------------
  
  Error with _naupd info =           -9
  Check the documentation of _naupd 
  
  
  Error with _nband, info=           -9
  Check the documentation of _nband 
  
<end of output>
Test time =   0.02 sec
----------------------------------------------------------

```

**Reply by wo80** [reply]
Alright. Thanks! Since I don't know about the _cmake non-reg test suite_, maybe someone who does know should do the PR and fix both the problem with that test suite and the actual problem with both example programs. What do you think?
(of course, I'm still willing to do the fix for the examples)

**Reply by fghoussen** [reply]
PR the fix. @sylvestre may add the cmake/autotools part on top of it before merge.

**Reply by fghoussen** [reply]
`call exit(1)` (do not use `stop` which returns 0) is missing here https://github.com/opencollab/arpack-ng/blob/569a3859c1bf79b303d830ada7b745de0f18512c/EXAMPLES/BAND/cnbdr2.f#L315 This is likely true for all examples.

**Reply by wo80** [reply]
> This is likely true for all examples.

I guess. So lets not clutter the PR with +100 `call exit(1)` additions, right?

**Reply by fghoussen** [reply]
Ideally one PR for each. Rebase the 2d (correction) on the 1st (exit call that will break CI)

**Reply by wo80** [reply]
I'll leave the PR as is. The exit code is probably a subject for a separate discussion and changes can be done after the merge. If you'd want to do it right, there should also be some check for the number of converged eigenvalues, like
```fortran
if ( nconv .lt. nev ) then
      call exit(1)
end if
```
